### ISSUE TITLE
oneOf not selecting correct default value for complex objects with the same property keys

### ISSUE THREAD

**Issue by Becaan** [rjsf-team/react-jsonschema-form#4208]
### Prerequisites

- [X] I have searched the existing issues
- [X] I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [X] I have read the [documentation](https://rjsf-team.github.io/react-jsonschema-form/docs)
- [X] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net), [Codesandbox.io](https://codesandbox.io) or preferably a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### What theme are you using?

core

### Version

5.x

### Current Behavior

As you can see in this example: https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJQbGFuSWQiOiIxMjMiLCJTdW0gSW5zdXJlZCI6IjkwMDAiLCJIZWFydCBhbmQgVmFzY3VsYXIgU3lzdGVtIjp7IlN1bSBJbnN1cmVkIjoiMTIwMCJ9fSwic2NoZW1hIjp7InR5cGUiOiJvYmplY3QiLCJvbmVPZiI6W3sidGl0bGUiOiJQbGFuIiwicHJvcGVydGllcyI6eyJQbGFuSWQiOnsidHlwZSI6InN0cmluZyIsImRlZmF1bHQiOiI2MSIsInJlYWRPbmx5Ijp0cnVlfSwiU3VtIEluc3VyZWQiOnsidHlwZSI6InN0cmluZyIsImRlZmF1bHQiOiIxMjAwIiwicGF0dGVybiI6Il5bMC05XXsxLDIwMn0oXFwuWzAtOV17MSwyM30pPyQifSwiSGVhcnQgYW5kIFZhc2N1bGFyIFN5c3RlbSI6eyJwcm9wZXJ0aWVzIjp7IlN1bSBJbnN1cmVkIjp7InR5cGUiOiJzdHJpbmciLCJkZWZhdWx0IjoiMTIwMCIsInBhdHRlcm4iOiJeWzAtOV17MSwyMDJ9KFxcLlswLTldezEsMjN9KT8kIn19fX19LHsidGl0bGUiOiJOZXcgUGxhbiIsInByb3BlcnRpZXMiOnsiUGxhbklkIjp7InR5cGUiOiJzdHJpbmciLCJkZWZhdWx0IjoiMTIzIiwicmVhZE9ubHkiOnRydWV9LCJTdW0gSW5zdXJlZCI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6IjkwMDAiLCJwYXR0ZXJuIjoiXlswLTldezEsMjAyfShcXC5bMC05XXsxLDIzfSk/JCJ9LCJIZWFydCBhbmQgVmFzY3VsYXIgU3lzdGVtIjp7InByb3BlcnRpZXMiOnsiU3VtIEluc3VyZWQiOnsidHlwZSI6InN0cmluZyIsImRlZmF1bHQiOiI5MDAwIiwicGF0dGVybiI6Il5bMC05XXsxLDIwMn0oXFwuWzAtOV17MSwyM30pPyQifX19fX1dfSwidWlTY2hlbWEiOnt9LCJ0aGVtZSI6Im1hdGVyaWFsLXVpLTUiLCJsaXZlU2V0dGluZ3MiOnt9fQ==

The default value for complex object is not taken into account at all as well as pattern for example which i tried also.
So basically. "Sum Insured" that is a part of "Heart and Vascular System" always has the same default value of 1200 even though in jsonSchema default value for "New Plan" is set to 9000.
The behaviour that it should follow should be like the "Sum Insured" that is outside of "Heart and Vascular System".

TLDR: "Sum Insured" inside the "Heart and Vascular System" should have default value 9000 previewed when selecting "New Plan"

### Expected Behavior

"Sum Insured" inside the "Heart and Vascular System" should have default value 9000 previewed when selecting "New Plan"

### Steps To Reproduce

https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJQbGFuSWQiOiI2MSIsIlN1bSBJbnN1cmVkIjoiMTIwMCIsIkhlYXJ0IGFuZCBWYXNjdWxhciBTeXN0ZW0iOnsiU3VtIEluc3VyZWQiOiIxMjAwIn19LCJzY2hlbWEiOnsidHlwZSI6Im9iamVjdCIsIm9uZU9mIjpbeyJ0aXRsZSI6IlBsYW4iLCJwcm9wZXJ0aWVzIjp7IlBsYW5JZCI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6IjYxIiwicmVhZE9ubHkiOnRydWV9LCJTdW0gSW5zdXJlZCI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6IjEyMDAiLCJwYXR0ZXJuIjoiXlswLTldezEsMjAyfShcXC5bMC05XXsxLDIzfSk/JCJ9LCJIZWFydCBhbmQgVmFzY3VsYXIgU3lzdGVtIjp7InByb3BlcnRpZXMiOnsiU3VtIEluc3VyZWQiOnsidHlwZSI6InN0cmluZyIsImRlZmF1bHQiOiIxMjAwIiwicGF0dGVybiI6Il5bMC05XXsxLDIwMn0oXFwuWzAtOV17MSwyM30pPyQifX19fX0seyJ0aXRsZSI6Ik5ldyBQbGFuIiwicHJvcGVydGllcyI6eyJQbGFuSWQiOnsidHlwZSI6InN0cmluZyIsImRlZmF1bHQiOiIxMjMiLCJyZWFkT25seSI6dHJ1ZX0sIlN1bSBJbnN1cmVkIjp7InR5cGUiOiJzdHJpbmciLCJkZWZhdWx0IjoiOTAwMCIsInBhdHRlcm4iOiJeWzAtOV17MSwyMDJ9KFxcLlswLTldezEsMjN9KT8kIn0sIkhlYXJ0IGFuZCBWYXNjdWxhciBTeXN0ZW0iOnsicHJvcGVydGllcyI6eyJTdW0gSW5zdXJlZCI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6IjkwMDAiLCJwYXR0ZXJuIjoiXlswLTldezEsMjAyfShcXC5bMC05XXsxLDIzfSk/JCJ9fX19fV19LCJ1aVNjaGVtYSI6e30sInRoZW1lIjoibWF0ZXJpYWwtdWktNSIsImxpdmVTZXR0aW5ncyI6e319

### Environment

```markdown
- OS:
- Node:
- npm:
```

**Comment by nickgros** [comment]
@Becaan thanks for the report, this is definitely a bug. The regular maintainers don't have much time to fix all issues, are you interested to investigate a fix?

This looks similar to #3892, which dealt with `allOf`s, which was fixed by #3969, so that PR may be a good place to figure out where to start

**Comment by Writingsinfo** [comment]
Hi @nickgros I know you have labeled it as **_help wanted_** but is it possible to escalate this to the team and see if anyone can pick it up? We are relying on oneOf and have been facing the same issue.

**Comment by Writingsinfo** [comment]
The issue we are seeing is that when we switch options, the default values are restored only for the fields that are common to both the options. Eg. [Link](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJjb21tb24iOiJjb21tb24ifSwic2NoZW1hIjp7InR5cGUiOiJvYmplY3QiLCJvbmVPZiI6W3sicHJvcGVydGllcyI6eyJsb3JlbSI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6ImxvcmVtIn0sImNvbW1vbiI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6ImNvbW1vbiJ9fSwicmVxdWlyZWQiOlsibG9yZW0iXX0seyJwcm9wZXJ0aWVzIjp7Imlwc3VtIjp7InR5cGUiOiJzdHJpbmciLCJkZWZhdWx0IjoiaXBzdW0ifSwiY29tbW9uIjp7InR5cGUiOiJzdHJpbmciLCJkZWZhdWx0IjoiY29tbW9uIn19LCJyZXF1aXJlZCI6WyJpcHN1bSJdfV19LCJ1aVNjaGVtYSI6e30sInRoZW1lIjoiZGVmYXVsdCIsImxpdmVTZXR0aW5ncyI6eyJzaG93RXJyb3JMaXN0IjoidG9wIiwiZXhwZXJpbWVudGFsX2RlZmF1bHRGb3JtU3RhdGVCZWhhdmlvciI6eyJhcnJheU1pbkl0ZW1zIjp7InBvcHVsYXRlIjoicG9wdWxhdGUiLCJtZXJnZUV4dHJhRGVmYXVsdHMiOmZhbHNlfSwiYWxsT2YiOiJza2lwRGVmYXVsdHMiLCJlbXB0eU9iamVjdEZpZWxkcyI6InBvcHVsYXRlQWxsRGVmYXVsdHMifX19)

**Comment by nickgros** [comment]
@Becaan We took another look at your schema and realized that the issue is that the `"Heart and Vascular System" property is missing `"type": "object"`. If you add that, it works ([playground link](https://rjsf-team.github.io/react-jsonschema-form/#eyJmb3JtRGF0YSI6eyJQbGFuSWQiOiIxMjMiLCJTdW0gSW5zdXJlZCI6IjkwMDAiLCJIZWFydCBhbmQgVmFzY3VsYXIgU3lzdGVtIjp7IlN1bSBJbnN1cmVkIjoiOTAwMCJ9fSwic2NoZW1hIjp7InR5cGUiOiJvYmplY3QiLCJvbmVPZiI6W3sidGl0bGUiOiJQbGFuIiwicHJvcGVydGllcyI6eyJQbGFuSWQiOnsidHlwZSI6InN0cmluZyIsImRlZmF1bHQiOiI2MSIsInJlYWRPbmx5Ijp0cnVlfSwiU3VtIEluc3VyZWQiOnsidHlwZSI6InN0cmluZyIsImRlZmF1bHQiOiIxMjAwIiwicGF0dGVybiI6Il5bMC05XXsxLDIwMn0oXFwuWzAtOV17MSwyM30pPyQifSwiSGVhcnQgYW5kIFZhc2N1bGFyIFN5c3RlbSI6eyJ0eXBlIjoib2JqZWN0IiwicHJvcGVydGllcyI6eyJTdW0gSW5zdXJlZCI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6IjEyMDAiLCJwYXR0ZXJuIjoiXlswLTldezEsMjAyfShcXC5bMC05XXsxLDIzfSk/JCJ9fX19fSx7InRpdGxlIjoiTmV3IFBsYW4iLCJwcm9wZXJ0aWVzIjp7IlBsYW5JZCI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6IjEyMyIsInJlYWRPbmx5Ijp0cnVlfSwiU3VtIEluc3VyZWQiOnsidHlwZSI6InN0cmluZyIsImRlZmF1bHQiOiI5MDAwIiwicGF0dGVybiI6Il5bMC05XXsxLDIwMn0oXFwuWzAtOV17MSwyM30pPyQifSwiSGVhcnQgYW5kIFZhc2N1bGFyIFN5c3RlbSI6eyJ0eXBlIjoib2JqZWN0IiwicHJvcGVydGllcyI6eyJTdW0gSW5zdXJlZCI6eyJ0eXBlIjoic3RyaW5nIiwiZGVmYXVsdCI6IjkwMDAiLCJwYXR0ZXJuIjoiXlswLTldezEsMjAyfShcXC5bMC05XXsxLDIzfSk/JCJ9fX19fV19LCJ1aVNjaGVtYSI6e30sInRoZW1lIjoibXVpIiwibGl2ZVNldHRpbmdzIjp7InNob3dFcnJvckxpc3QiOiJ0b3AiLCJleHBlcmltZW50YWxfZGVmYXVsdEZvcm1TdGF0ZUJlaGF2aW9yIjp7ImFycmF5TWluSXRlbXMiOnsicG9wdWxhdGUiOiJwb3B1bGF0ZSIsIm1lcmdlRXh0cmFEZWZhdWx0cyI6ZmFsc2V9LCJhbGxPZiI6InNraXBEZWZhdWx0cyIsImVtcHR5T2JqZWN0RmllbGRzIjoicG9wdWxhdGVBbGxEZWZhdWx0cyJ9fX0=))

**Comment by nickgros** [comment]
@Writingsinfo @heath-freenome is going to have a PR soon that will hopefully fix your issue in #4322
